### PR TITLE
Respect Git tags and branches that look like short commits

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

If we're given a Git reference like `20240222`, we currently treat it as a short commit hash. However... it _could_ be a branch or a tag. This PR improves the Git reference logic to ensure that ambiguous references like `20240222` are handled appropriately, by attempting to extract it as a branch, then a tag, then a short commit hash.

Closes https://github.com/astral-sh/uv/issues/2772.

### CODE REVIEW

**Comment by charliermarsh** [comment]
(Needs tests.)

File: crates/uv-git/src/git.rs [hunk 1]
```diff
@@ -25,20 +25,14 @@ const CHECKOUT_READY_LOCK: &str = ".ok";
 /// A reference to commit or commit-ish.
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
 pub enum GitReference {
-    /// From a branch.
-    #[allow(unused)]
-    Branch(String),
-    /// From a tag.
-    #[allow(unused)]
-    Tag(String),
```
Comment (charliermarsh) [inline crates/uv-git/src/git.rs:33]: These were removed. (See individual commits.)

File: crates/uv-git/src/git.rs [hunk 2]
```diff
@@ -25,20 +25,14 @@ const CHECKOUT_READY_LOCK: &str = ".ok";
...
     /// From a reference that's ambiguously a branch or tag.
     BranchOrTag(String),
+    /// From a reference that's ambiguously a short commit, a branch, or a tag.
+    BranchOrTagOrCommit(String),
+    /// From a named reference, like `refs/pull/493/head`.
+    NamedRef(String),
```
Comment (charliermarsh) [inline crates/uv-git/src/git.rs:33]: This was renamed from `Ref`.

File: crates/uv-git/src/git.rs [hunk 3]
```diff
@@ -25,20 +25,14 @@ const CHECKOUT_READY_LOCK: &str = ".ok";
...
     /// From a specific revision, using a full 40-character commit hash.
     FullCommit(String),
-    /// From a truncated revision.
-    ShortCommit(String),
```
Comment (charliermarsh) [inline crates/uv-git/src/git.rs:39]: This is now `BranchOrTagOrCommit`.

**Comment by zanieb** [comment]
I'll review this tonight.

**Review by BurntSushi — APPROVED** [review]
Looks reasonable to me. I liked the commit breakdown. :)

**Review by zanieb — APPROVED** [review]

**Pull request merged**